File: leetcode/go/trees/binaryTreeLevelOrderTraversal(medium)/binaryTreeLevelOrderTraversal.go
package main

import "fmt"

/* 102. Binary Tree Level Order Traversal
https://leetcode.com/problems/binary-tree-level-order-traversal/description/

Учитывая корень двоичного дерева, вернуть порядок обхода значений его узлов (т. е. слева направо, уровень за уровнем).

Example 1:
Input: root = [3,9,20,null,null,15,7]
Output: [[3],[9,20],[15,7]]

Example 2:
Input: root = [1]
Output: [[1]]
*/

func main() {
	root := createTree([]interface{}{3, 9, 20, nil, nil, 15, 7})
	fmt.Println(levelOrder(root)) // [[3],[9,20],[15,7]]
}

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// levelOrder возвращает список значений узлов дерева в порядке обхода уровнями.
// time: O(n), space: O(n), где n - количество узлов в дереве
func levelOrder(root *TreeNode) [][]int {
	levels := make([][]int, 0)
	preOrder(root, 0, &levels)
	return levels
}

// preOrder рекурсивно обходит дерево и заполняет уровни.
// time: O(n), space: O(n)
func preOrder(node *TreeNode, level int, levels *[][]int) *[][]int {
	// Если узел равен nil, возвращаем слайс уровней
	if node == nil {
		return levels
	}

	// Если текущего уровня еще нет в слайсе, добавляем новый уровень
	if level == len(*levels) {
		*levels = append(*levels, []int{})
	}

	// Делаем PreOrder обход дерева и добавляем вершину на текущий уровень в конец списка.
	(*levels)[level] = append((*levels)[level], node.Val)
	// Рекурсивно обходим левое и правое поддеревья
	preOrder(node.Left, level+1, levels)
	preOrder(node.Right, level+1, levels)

	return levels
}

// createTree создает бинарное дерево из массива значений.
func createTree(values []interface{}) *TreeNode {
	if len(values) == 0 || values[0] == nil {
		return nil
	}

	root := &TreeNode{Val: values[0].(int)} // Создаем корневой узел с первым значением в массиве
	queue := []*TreeNode{root}              // Создаем очередь для обхода дерева по уровням
	i := 1                                  // Индекс для обхода массива значений

	for len(queue) > 0 && i < len(values) {
		node := queue[0]  // Берем первый элемент из очереди
		queue = queue[1:] // Удаляем первый элемент из очереди

		// Левый потомок
		if i < len(values) && values[i] != nil {
			node.Left = &TreeNode{Val: values[i].(int)} // Создаем левый потомок и добавляем его в очередь
			queue = append(queue, node.Left)            // Добавляем левого потомка в очередь
		}
		i++

		// Правый потомок
		// Если правый потомок не равен nil, то создаем его и добавляем его в очередь
		if i < len(values) && values[i] != nil {
			node.Right = &TreeNode{Val: values[i].(int)} // Создаем правого потомка и добавляем его в очередь
			queue = append(queue, node.Right)            // Добавляем правого потомка в очередь
		}
		i++
	}

	// Возвращаем указатель на корневой узел
	return root
}

File: leetcode/java/trees/BinaryTreeLevelOrderTraversal/BinaryTreeLevelOrderTraversal.java
package leetcode.java.trees.BinaryTreeLevelOrderTraversal;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/* 102. Binary Tree Level Order Traversal
https://leetcode.com/problems/binary-tree-level-order-traversal/description/

Учитывая корень двоичного дерева, вернуть порядок обхода значений его узлов (т. е. слева направо, уровень за уровнем).

Example 1:
Input: root = [3,9,20,null,null,15,7]
Output: [[3],[9,20],[15,7]]

Example 2:
Input: root = [1]
Output: [[1]]
*/

public class BinaryTreeLevelOrderTraversal {
    public static void main(String[] args) {
        TreeNode root = createTree(new Integer[]{3, 9, 20, null, null, 15, 7});
        System.out.println(levelOrder(root)); // [[3], [9, 20], [15, 7]] 
    }

    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int val) {
            this.val = val;
        }
    }

    // levelOrder возвращает список значений узлов дерева в порядке обхода уровнями.
    // time: O(n), space: O(n), где n - количество узлов в дереве
    private static List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> levels = new ArrayList<>();
        preOrder(root, 0, levels);
        return levels;
    }

    // preOrder рекурсивно обходит дерево и заполняет уровни.
    // time: O(n), space: O(n)
    private static void preOrder(TreeNode node, int level, List<List<Integer>> levels) {
        // Если узел равен null, возвращаемся
        if (node == null) {
            return;
        }

        // Если текущего уровня еще нет в списке, добавляем новый уровень
        if (level == levels.size()) {
            levels.add(new ArrayList<>());
        }

        // Делаем PreOrder обход дерева и добавляем вершину на текущий уровень в конец списка.
        levels.get(level).add(node.val);
        // Рекурсивно обходим левое и правое поддеревья
        preOrder(node.left, level + 1, levels);
        preOrder(node.right, level + 1, levels);
    }

    // createTree создает бинарное дерево из массива значений.
    private static TreeNode createTree(Integer[] values) {
        if (values == null || values.length == 0 || values[0] == null) {
            return null;
        }

        TreeNode root = new TreeNode(values[0]); // Создаем корневой узел с первым значением в массиве
        Queue<TreeNode> queue = new LinkedList<>(); // Создаем очередь для обхода дерева по уровням
        queue.offer(root);
        int i = 1; // Индекс для обхода массива значений

        while (!queue.isEmpty() && i < values.length) {
            TreeNode node = queue.poll(); // Берем первый элемент из очереди

            // Левый потомок
            if (i < values.length && values[i] != null) {
                node.left = new TreeNode(values[i]); // Создаем левый потомок и добавляем его в очередь
                queue.offer(node.left); // Добавляем левого потомка в очередь
            }
            i++;

            // Правый потомок
            // Если правый потомок не равен null, то создаем его и добавляем его в очередь
            if (i < values.length && values[i] != null) {
                node.right = new TreeNode(values[i]); // Создаем правого потомка и добавляем его в очередь
                queue.offer(node.right); // Добавляем правого потомка в очередь
            }
            i++;
        }

        // Возвращаем ссылку на корневой узел
        return root;
    }
}
